engine/buildkit/executor_spec.go

```diff
@@ -626,114 +616,136 @@ func (w *Worker) setupStdio(_ context.Context, state *execState) error {
-			return net.Listen("tcp", "127.0.0.1:0")
-		})
-		if err != nil {
-			return fmt.Errorf("otel tcp proxy listen: %w", err)
-		}
-		state.otelEndpoint = "http://" + listener.Addr().String()
+	otelProto := "http/protobuf"
+	otelEndpoint := "http://" + listener.Addr().String()
+	otelSrv := &http.Server{
+		Handler: w.telemetryPubSub,
 
-		proxyConnPool := pool.New()
-		listenerCtx, cancelListener := context.WithCancel(ctx)
-		listenerPool := pool.New().WithContext(listenerCtx).WithCancelOnError()
+		ReadHeaderTimeout: 10 * time.Second, // for gocritic
+	}
 
-		listenerPool.Go(func(ctx context.Context) error {
-			<-ctx.Done()
-			err := listener.Close()
+	listenerPool.Go(func(ctx context.Context) error {
+		<-ctx.Done()
+		err := listener.Close()
+		if err != nil {
+			return fmt.Errorf("close otel proxy listener: %w", err)
+		}
+		return nil
+	})
+	listenerPool.Go(func(ctx context.Context) error {
+		for {
+			conn, err := listener.Accept()
 			if err != nil {
-				return fmt.Errorf("close otel proxy listener: %w", err)
-			}
-			return nil
-		})
-		listenerPool.Go(func(ctx context.Context) error {
-			for {
-				conn, err := listener.Accept()
-				if err != nil {
-					if errors.Is(err, net.ErrClosed) {
-						err = nil
-					}
-					return err
-				}
-
-				// TODO:(sipsma) logging that existed before? Was it useful?
-
-				remote, err := net.Dial("unix", state.otelUnixSockSrcPath)
-				if err != nil {
-					conn.Close()
-					return fmt.Errorf("dial otel unix socket: %w", err)
+				if errors.Is(err, net.ErrClosed) {
+					err = nil
 				}
+				return err
+			}
 
-				proxyConnPool.Go(func() {
-					defer remote.Close()
-					io.Copy(remote, conn)
-				})
-
-				proxyConnPool.Go(func() {
-					defer conn.Close()
-					io.Copy(conn, remote)
-				})
+			l := &singleConnListener{
+				conn:    conn,
+				closeCh: make(chan struct{}),
 			}
-		})
-		state.cleanups.addNoErr("wait for otel proxy conn pool", proxyConnPool.Wait)
-		state.cleanups.add("wait for otel listener pool", listenerPool.Wait)
-		state.cleanups.addNoErr("cancel otel listener context", cancelListener)
-	}
+			proxyConnPool.Go(func() {
+				<-listenerCtx.Done()
+				l.Close()
+			})
 
+			proxyConnPool.Go(func() {
+				otelSrv.Serve(l)
+			})
+		}
+	})
+	state.cleanups.addNoErr("wait for otel proxy conn pool", proxyConnPool.Wait)
+	state.cleanups.add("wait for otel listener pool", listenerPool.Wait)
+	state.cleanups.addNoErr("cancel otel listener context", cancelListener)
+
+	// Configure our OpenTelemetry proxy. A lot.
 	state.spec.Process.Env = append(state.spec.Process.Env,
-		OTELExporterProtocolEnv+"="+state.otelProto,
-		OTELExporterEndpointEnv+"="+state.otelEndpoint,
-		OTELTracesProtocolEnv+"="+state.otelProto,
-		OTELTracesEndpointEnv+"="+state.otelEndpoint,
+		OTelExporterProtocolEnv+"="+otelProto,
+		OTelExporterEndpointEnv+"="+otelEndpoint,
+		OTelTracesProtocolEnv+"="+otelProto,
+		OTelTracesEndpointEnv+"="+otelEndpoint+"/v1/traces",
```

Comment:
Pretty sure this var along with the log + metrics equivalent were just wrong before - I was getting `404` errors because telemetry was being sent to the base URL, without `/v1/(traces|logs|metrics)`.